search/searcher/search_conjunction.go

```diff
@@ -205,12 +220,40 @@ OUTER:
 			// don't bump i, so that we'll examine the just-advanced
 			// currs[i] again
 		}
+		// if we get here, a doc matched all readers
+		if s.searcherPositions == nil {
+			rv = s.scorer.Score(ctx, s.currs)
+		} else {
+			// we only score the document if the relative positioning rule is satisfied
+			// if not, we advance the maxIDIdx and try again
+			var hitTable = make([][]PositionPair, len(s.currs))
+			for id, match := range s.currs {
+				hitTable[id] = make([]PositionPair, len(match.FieldTermLocations))
+				firstPos := match.FieldTermLocations[0].Location.Pos
+				lastPos := firstPos
+				colID := 0
+				for i := 1; i < len(match.FieldTermLocations); i++ {
+					if match.FieldTermLocations[i].QueryIndex > match.FieldTermLocations[i-1].QueryIndex ||
+						match.FieldTermLocations[i].OccurrenceIndex > match.FieldTermLocations[i-1].OccurrenceIndex {
+						hitTable[id][colID] = PositionPair{firstPos, lastPos}
+						colID++
+						firstPos = match.FieldTermLocations[i].Location.Pos
+						lastPos = firstPos
+					} else {
+						lastPos = match.FieldTermLocations[i].Location.Pos
+					}
+				}
+				hitTable[id][colID] = PositionPair{firstPos, lastPos}
+				hitTable[id] = hitTable[id][:colID+1]
+			}
 
-		// if we get here, a doc matched all readers, so score and add it
-		rv = s.scorer.Score(ctx, s.currs)
-
-		// we know all the searchers are pointing at the same thing
-		// so they all need to be bumped
+			for i := 0; i < len(hitTable[0]); i++ {
+				if positionsAreCorrect(s.searcherPositions, 1, hitTable, hitTable[0][i].LastPos) {
```

Comment:
since this dfs algorithm is called for every hit, there will be perf implications. so i think its better to optimise it by storing a flag which says whether the path from this particular hitTable entry will actually match the phrase or not. this way, the next iterations would simply check the flag before executing the dfs along that same identical path